### PR TITLE
Add Hugging Face relevance training script

### PR DESCRIPTION
**Pull request by EvilDrPurple**
#### Fixes

* #31 

#### Description

* Adds current training script for relevance model, predicts if a url and its contents are either relevant or irrelevant

### CODE REVIEW

**Comment by mbodeantor** [comment]
@EvilDrPurple Getting this error when I try to run huggingfac_relevance.py:
RuntimeError: The size of tensor a (3598) must match the size of tensor b (2048) at non-singleton dimension 1

**Comment by mbodeantor** [comment]
@EvilDrPurple also can you add multimodal_transformers to requirements.txt? also upload one of the files you used with clean_data.py?

**Comment by EvilDrPurple** [comment]
> @EvilDrPurple Getting this error when I try to run huggingfac_relevance.py: RuntimeError: The size of tensor a (3598) must match the size of tensor b (2048) at non-singleton dimension 1

@mbodeantor I went ahead and updated the requirements.txt. I could not reproduce the error in a fresh venv. If it still doesn't work would you mind using `pip list` and sending it my way so that I may compare with my package versions?

**Comment by EvilDrPurple** [comment]
Additional notes about this:

I was having issues with `Python 3.12` as well as newer versions of `pytest` and `tqdm` since `multimodal-transformers` requires some older versions of those packages. I downgraded them in the `requirements.txt` but if this causes problems with other scripts we may need to split the requirements and have a more localized one. This script will effectively require `Python 3.11` or lower to function until certain packages are updated to support `3.12`

**Comment by josh-chamberlain** [comment]
@maxachis could you please pick up the review on this one?

**Comment by maxachis** [comment]
@EvilDrPurple Here I am, picking up the review! 

I ran into the following error when installing the requirements, specifically when I got to 
`Collecting scikit-learn~=1.0.2 (from multimodal-transformers>=0.3.1->-r requirements.txt (line 21))`

Here is the [error.txt](https://github.com/Police-Data-Accessibility-Project/data-source-identification/files/15358056/error.txt). There was more to it than that, but this seemed the most directly relevant part. Here, additionally, is the 
[pip_list.txt](https://github.com/Police-Data-Accessibility-Project/data-source-identification/files/15358063/pip_list.txt).

I got this error under two circumstances: One while attempting to run this on my personal computer, and one while running a Dockerfile built using the following instructions (which I'd separately developed to run some of my own tests): [GithubTest.zip](https://github.com/Police-Data-Accessibility-Project/data-source-identification/files/15358031/GithubTest.zip). So it's _probably_ not a 'works on my machine' problem. 

Could you have a look at this error (perhaps using the same dockerfile as I did) and see what's going on?

**Comment by EvilDrPurple** [comment]
@maxachis After a good bit of trial and error I figured out the issue. Basically `multimodal-transformers` requires an older version of `scikit-learn` which cannot be installed using newer versions of `Cython` bundled in more recent Python releases. This effectively means that installing from requirements.txt requires `Python 3.10` or lower to function without a workaround. I added a README to reflect this requirement and provide instructions for setting up the environment. Let me know if you are able to run on your end using a 3.10 instance.

**Comment by maxachis** [comment]
> @maxachis After a good bit of trial and error I figured out the issue. Basically `multimodal-transformers` requires an older version of `scikit-learn` which cannot be installed using newer versions of `Cython` bundled in more recent Python releases. This effectively means that installing from requirements.txt requires `Python 3.10` or lower to function without a workaround. I added a README to reflect this requirement and provide instructions for setting up the environment. Let me know if you are able to run on your end using a 3.10 instance.

@EvilDrPurple Got it! I intend to get to it tomorrow as my first bit of PDAP work, but if I don't, please message me about it!

**Comment by maxachis** [comment]
@EvilDrPurple Downloading the various requirements right now. One request I'll make (as a [nit](https://google.github.io/eng-practices/review/reviewer/standard.html)) is to try breaking down the requirements so that the requirements for this module specifically exist within the module. Otherwise, we're installing all the modules for the repository, including ones not relevant to this module.

Note that this is in contradiction with a much earlier refactoring I did, where I tried to consolidate the requirements into a single requirements.txt 🙃. Lesson learned on my part.

**Comment by maxachis** [comment]
@EvilDrPurple I'm able to get it running successfully and perform training! However, as to whether that training _works_, I don't know. I'm willing to tentatively approve this but will defer to you and/or @josh-chamberlain if you feel it requires additional testing. 

Additionally, I recommend that in the future you have the PR conform to [this](https://github.com/Police-Data-Accessibility-Project/docs/blob/main/meta/operations/staff/github-pull-request-template.md) PR template, just because that makes it easier for me to run it and to know what to look for.

File: hugging_face/url_relevance/clean_data.py
Comment (maxachis) [inline hugging_face/url_relevance/clean_data.py:1]: Add a descriptive docstring in the heading of this file to better describe what it's doing and how it relates to everything else.

File: hugging_face/url_relevance/huggingface_relevance.py
Comment (maxachis) [inline hugging_face/url_relevance/huggingface_relevance.py:1]: For this as well, add a descriptive docstring.

File: hugging_face/example/clean_data.py
Comment (maxachis) [inline hugging_face/example/clean_data.py:1]: Please add a descriptive docstring for this file to describe what it does and how it relates to the rest of the functionality.

**Review by maxachis — CHANGES_REQUESTED** [review]

**Comment by EvilDrPurple** [comment]
@maxachis docstring added, I also renamed one the `clean_data.py` scripts to `split_data.py` because I think that described better what is happening

**Review by maxachis — APPROVED** [review]
Looks good! Several nits:

1. Try to apply black so that the linter doesn't get irritated in the Github Action
2. You have some commented-out code. If you don't need it, recommend getting rid of it (as someone else will end up doing the same thing later anyway).
3. Rather than having clauses several indentations deep, recommend extracting logic into separate functions -- this can also make things more self-documenting.

**Pull request merged**